### ISSUE TITLE
Cannot decrypt messages from specific user sent using specific device: `Unable to decrypt: OLM`

### ISSUE THREAD

**Issue by calvinrw** [vector-im/element-android#3429]
#### Describe the bug

When using Element Android on a phone running Android 10, E2E messages received from one specific device belonging to another user are not immediately viewable. Instead, `Waiting for this message, this may take a while` is seen in place of the message contents, or `** Unable to decrypt: OLM **` when Developer mode is turned on in Element's advanced settings.

#### To Reproduce

A room exists consisting of me and my friend Ed, each using our own, self-hosted Synapse homeservers. We have not tested for this issue in other rooms.

1. Ed uses his phone to send a message to this room.
2. I observe the following:
   a. On my phone running Android 10: Often `Unable to decrypt: OLM` is shown in place of the message contents. Otherwise, the message contents are visible, as expected.
   b. On my phone running Android 6: The message contents are visible, as expected.
   c. On my computer with Element Desktop: The message contents visible, as expected.

For comparison, we tested the following:

1. Ed uses his computer with Element Desktop to send a message to this room.
2. I observe the following:
   a. On my phone running Android 10: The message contents visible, as expected.
   b. On my phone running Android 6: The message contents are visible, as expected.
   c. On my computer with Element Desktop: The message contents are visible, as expected.
   
We also tested:

1. I use my Android 10 phone to send a message to this room.
2. On Ed's phone: The message contents are visible, as expected.

Finally, we tested:

1. In a new room, without E2E enabled, Ed uses his phone to send a message to the room.
2. I observe the following:
   a. On my phone running Android 10: The message contents are visible, as expected.
   b. On my phone running Android 6: The message contents are visible, as expected.
   c. On my computer with Element Desktop: The message contents are visible, as expected.
     
Each test consisted of Ed (or myself) sending multiple messages in quick succession, and the 'failure' rate for my ability to view messages on the affected phone was 100%. But under normal, "real life" circumstances (i.e. an actual conversation), this only occurred once every few messages.

I've concluded that this issue only affects messages sent from Ed's phone when they're viewed on my Android 10 phone, and in a room with E2E enabled.

On the affected (Android 10) phone, logs were captured using `logcat` over `adb`. The following is shown repeatedly when affected messages are received:

```
05-24 21:55:02.212 27914  1430 E OlmJniNative:  ## decryptMessageJni(): failure - olm_group_decrypt Msg=BAD_MESSAGE_MAC
05-24 21:55:02.212 27914  1430 E OlmInboundGroupSession: ## decryptMessage() failed BAD_MESSAGE_MAC
```

The affected messages eventually become visible on the Android 10 phone, but I have not figured out how long the delay actually is. It appears to be greater than a few hours.

#### Expected behavior

When Ed sends me a message, it will be viewable on every device I have Element and E2E correctly set up on in a reasonable amount of time, regardless of which of his devices he uses to send it.

#### Screenshots

Screenshots have been attached.

##### My phone affected by this issue

- Device: Motorola Moto Z3 Play
- OS: Android 10 (LineageOS 17.1, official, build 17.1-20210303-NIGHTLY-beckham)
 
##### My phone unaffected by this issue

- Device: Motorola Moto X
- OS: Android 6 (LineageOS 14, unofficial, self-built)

##### Ed's phone

- Device: Google Pixel 3
- OS: Android 11

#### Additional context

- Using Element Android: 1.1.7 via Google Play (Applies to all phones listed above)
- Homeserver: matrix.calvinrw.com (Synapse 1.34)
- Ed's homeserver: Synapse 1.34
- Two rageshakes were submitted on 2021-05-23
- This issue did not occur prior to 2021-05-05, and started occurring consistently after 2021-05-10.
- E2E appears to be set up correctly on all devices, because Element shows the green shield for this room. We have verified each other, and all of our sessions are verified.
- Before and during the tests described above, both of my phones were kept awake using LineageOS's caffeine feature, so Android's power management was not a factor.

![Screenshot_Android6_20210523-134331_resize](https://user-images.githubusercontent.com/687129/120088961-9c61bb00-c0bb-11eb-9a45-4c878768f753.png)
![Screenshot_Android10_20210523-134337_Element_resize](https://user-images.githubusercontent.com/687129/120088962-9c61bb00-c0bb-11eb-952f-bd62a9cd4033.png)

**Comment by kescherCode** [comment]
I seem to have this issue, where I am always one message behind in comparison to desktop in the only room this happens in. As soon as the other party sends another encrypted messages, the "Unable to decrypt: OLM" becomes visible, but that next message becomes "Unable to decrypt: OLM" once more.

**Comment by jeffw387** [comment]
I was having this issue, I managed to solve it for now by signing out on the device then back in, and re-verifying the session.

**Comment by kescherCode** [comment]
@jeffw387 yes, but I don't want to relog every single time this issue occurs. This hinders the "encrypt only to verified sessions" feature.

**Comment by ewsandor** [comment]
Ed here from the original issue report.  

We are now seeing the same issue in reverse.  I am unable to see calvinrw's messages on Android with the same error message, but the same message comes through okay on the Element web client.

We are both now on Android Element version 1.1.9.

**Comment by skylarkbe** [comment]
Same kind of story over here. Where it becomes completely strange is when I receive the message properly on my desktop, on my Samsung Android tablet, but not on my Android phone, despite reverifying the session. When a new message arrives, I have to go back to the conversation list, reopen it and the missing messages are displayed ... except for the latest.

Could it be linked to https://github.com/vector-im/element-android/issues/1721 ?

**Comment by meyerrj** [comment]
This is happening for me as well. Pixel 4a, Android 12. 

Long-pressing and selecting "Re-request encryption keys from your other sessions." does not work. I have it open on two desktops and an iPad and it never decrypts.

**Comment by BillCarsonFr** [comment]
A new RS https://github.com/matrix-org/element-android-rageshakes/issues/27806

**Comment by HarHarLinks** [comment]
I just submitted a rageshake for (probably) this issue

**Comment by 5HT2** [comment]
Just submitted a rageshake for this exact issue. Messages from a user work fine when sent from their desktop, `Unable to decrypt: OLM` from their phone. Only happens on my Pixel 2 / Android 12, not reproducible on desktop recipients.

**Comment by hexxone** [comment]
I just also had the Problem a 2nd time. After the first on, I did a re-log, which fixxed it for about a week.
Now `Unable to decrypt: OLM` randomly starts appearing again.
Is there any progress on this issue?

**Comment by toshanmugaraj** [comment]
yes this issue happens in iOS too `olm_decrypt error: BAD_MESSAGE_MAC`

**Comment by mcg-matrix** [comment]
Just happened to me for the first time, in a 1:1 room. I have several sessions for that Matrix account, and only one of the sessions was affected. The affected session is currently running Element-FDroid 1.5.30.

Restarting the affected Element-FDroid did not help.

Executing "Settings" "Security & Privacy" "Encrypted Messages Recovery" "RESTORE FROM BACKUP" did help. I presume that at least one of the other sessions had stored the missing secret into the central backup. So for anyone using this feature (that used to be called "secure backup", I believe), this might be a more feasible approach than logging out and back in.

**Comment by ithinuel** [comment]
Happening to me now.
Messages sent from the desktop app cannot be received by (unverified) others (eg new encrypted rooms).
Older channels seem to still work (with a verified peer).

I will try to sign out/sign in from this app and report if this helps or not.
